doc/rascsi.1
.TH rascsi 1
.SH NAME
rascsi \- Emulates SCSI devices using the Raspberry Pi GPIO pins
.SH SYNOPSIS
.B rascsi
[\fB\-F\fR \fIFOLDER\fR]
[\fB\-L\fR \fILOG_LEVEL[:ID:[LUN]]\fR]
[\fB\-P\fR \fIACCESS_TOKEN_FILE\fR]
[\fB\-R\fR \fISCAN_DEPTH\fR]
[\fB\-h\fR]
[\fB\-n\fR \fIVENDOR:PRODUCT:REVISION\fR]
[\fB\-p\fR \fIPORT\fR]
[\fB\-r\fR \fIRESERVED_IDS\fR]
[\fB\-n\fR \fITYPE\fR]
[\fB\-v\fR]
[\fB\-z\fR \fILOCALE\fR]
[\fB\-IDn:[u]\fR \fIFILE\fR]
[\fB\-HDn[:u]\fR \fIFILE\fR]...
.SH DESCRIPTION
.B rascsi
emulates SCSI devices using the Raspberry Pi GPIO pins.
.PP
In the arguments to RaSCSI, one or more SCSI (-IDn[:u]) devices can be specified.
The number (n) after the ID or HD identifier specifies the ID number for that device. The optional number (u) specifies the LUN (logical unit) for that device. The default LUN is 0.
For SCSI: The ID is limited from 0-7. However, typically SCSI ID 7 is reserved for the "initiator" (the host computer). The LUN is limited from 0-31.
.PP
RaSCSI will determine the type of device based upon the file extension of the FILE argument.
    hd1: SCSI Hard Disk image (generic, non-removable, SCSI-1)
    hds: SCSI Hard Disk image (generic, non-removable)
    hdr: SCSI Hard Disk image (generic, removable)
    hdn: SCSI Hard Disk image (NEC compatible - only used with PC-98 computers)
    hdi: SCSI Hard Disk image (Anex86 proprietary - only used with PC-98 computers)
    nhd: SCSI Hard Disk image (T98Next proprietary - only used with PC-98 computers)
    hda: SCSI Hard Disk image (Apple compatible - typically used with Macintosh computers)
    mos: SCSI Magneto-Optical image (generic - typically used with NeXT, X68000, etc.)
    iso: SCSI CD-ROM or DVD-ROM image (ISO 9660 image)
  
For example, if you want to specify an Apple-compatible HD image on ID 0, you can use the following command:
    sudo rascsi -ID0 /path/to/drive/hdimage.hda

Once RaSCSI starts, it will open a socket (default port is 6868) to allow external management commands.
If another process is using the rascsi port, RaSCSI will terminate, since it is likely another instance of RaSCSI.
Once RaSCSI has initialized, the rasctl utility can be used to send commands.

To quit RaSCSI, press Control + C. If it is running in the background, you can kill it using an INT signal.

.SH OPTIONS
.TP
.BR \-b\fI " " \fIBLOCK_SIZE
The optional block size, either 512, 1024, 2048 or 4096 bytes. Default size is 512 bytes.
.TP
.BR \-F\fI " " \fIFOLDER
The default folder for image files. For files in this folder no absolute path needs to be specified. The initial default folder is '~/images'.
.TP
.BR \-L\fI " " \fILOG_LEVEL[:ID:[LUN]]
The rascsi log level (trace, debug, info, warn, err, off). The default log level is 'info' for all devices unless a particular device ID and an optional LUN was provided.
.TP
.BR \-P\fI " " \fIACCESS_TOKEN_FILE
Enable authentication and read the access token from the specified file. The access token file must be owned by root and must be readable by root only.
.TP
.BR \-R\fI " " \fISCAN_DEPTH
Scan for image files recursively, up to a depth of SCAN_DEPTH. Depth 0 means to ignore any folders within the default image filder. Be careful when using this option with many sub-folders in the default image folder. The default depth is 1.
.TP
.BR \-h\fI " " \fI
Show a help page.
.TP
.BR \-n\fI " " \fIVENDOR:PRODUCT:REVISION
Set the vendor, product and revision for the device, to be returned with the INQUIRY data. A complete set of name components must be provided. VENDOR may have up to 8, PRODUCT up to 16, REVISION up to 4 characters. Padding with blanks to the maxium length is automatically applied. Once set the name of a device cannot be changed.
.TP
.BR \-p\fI " " \fIPORT
The rascsi server port, default is 6868.
.TP
.BR \-r\fI " " \fIRESERVED_IDS
Comma-separated list of IDs to reserve. Pass an empty list in order to not reserve anything.
.BR \-p\fI " " \fITYPE
The optional case-insensitive device type (SAHD, SCHD, SCRM, SCCD, SCMO, SCBR, SCDP, SCLP, SCHS). If no type is specified for devices that support an image file, rascsi tries to derive the type from the file extension.
.TP
.BR \-v\fI " " \fI
Display the rascsi version.
.TP
.BR \-z\fI " "\fILOCALE
Overrides the default locale for client-faces error messages. The client can override the locale.
.TP
.BR \-ID\fIn[:u] " " \fIFILE
n is the SCSI ID number (0-7). u (0-31) is the optional LUN (logical unit). The default LUN is 0.
.IP
FILE is the name of the image file to use for the SCSI device. For devices that do not support an image file (SCBR, SCDP, SCLP, SCHS) the filename may have a special meaning or a dummy name can be provided. For SCBR and SCDP it is an optioinal prioritized list of network interfaces, an optional IP address and netmask, e.g. "interface=eth0,eth1,wlan0:inet=10.10.20.1/24". For SCLP it is the print command to be used and a reservation timeout in seconds, e.g. "cmd=lp -oraw %f:timeout=60".
.IP
FILE is the name of the image file to use for the SCSI device.
.IP

.SH EXAMPLES
Launch RaSCSI with no emulated drives attached:
   rascsi

Launch RaSCSI with an Apple hard drive image as ID 0 and a CD-ROM as ID 2
   rascsi -ID0 /path/to/harddrive.hda -ID2 /path/to/cdimage.iso

Launch RaSCSI with a removable SCSI drive image as ID 0 and the raw device file /dev/hdb (e.g. a USB stick) and a DaynaPort network adapter as ID 6:
   rascsi -ID0 -t scrm /dev/hdb -ID6 -t scdp daynaport

To create an empty, 100MiB HD image, use the following command:
   dd if=/dev/zero of=/path/to/newimage.hda bs=512 count=204800

In case the fallocate command is available a much faster alternative to the dd command is:
   fallocate -l 104857600 /path/to/newimage.hda

.SH SEE ALSO
rasctl(1), scsimon(1), rasdump(1)
 
Full documentation is available at: <https://www.github.com/akuker/RASCSI/wiki/>


src/php/rascsi.php
<!--  PHP source code for controlling the RaSCSI - 68kmla edition with a web interface. -->
<!--  Copyright (c) 2020 akuker -->
<!--  Distributed under the BSD-3 Clause License -->

<!-- Note: the origina rascsi-php project was under MIT license.-->

<!DOCTYPE html>
<html>

<head>
   <title>RaSCSI Main Control Page</title>
   <link rel="stylesheet" href="rascsi_styles.css">
 </head>

<body>

    <?php

	include 'lib_rascsi.php';
	html_generate_header();
	current_rascsi_config();
?>

    <br>
    <h2>Image File Management</h2>
    <table border="black" cellpadding="3">
        <tr>
            <td><b>Location</b></td>
            <td><b>Filename</b></td>
            <td><b>Size</b></td>
            <td><b>Type</b></td>
            <td><b>Date Modified</b></td>
            <td><b>Actions</b></td>
        </tr>
        <tr>
            <?php
	// Generate the table with all of the file names in it.....
	$all_files = get_all_files();
	foreach(explode(PHP_EOL, $all_files) as $this_file){
		if(strpos($this_file, 'total') === 0){
			continue;
		}
		$file_name = file_name_from_ls($this_file);
		if(strlen($file_name) === 0){
			continue;
		}
		// Ignore file names that start with .
		if(strpos($file_name,".") === 0){
			continue;
		}


		echo '<tr>'.PHP_EOL;
        echo '    <td>SD Card</td>'.PHP_EOL;
        echo '    <td>'.$file_name.'</td>'.PHP_EOL;
        echo '    <td>'.file_size_from_ls($this_file).'</td>'.PHP_EOL;
        echo '    <td>'.file_category_from_file_name($file_name).'</td>'.PHP_EOL;
        echo '    <td>'.mod_date_from_ls($this_file).'</td>'.PHP_EOL;
        echo '    <td>'.PHP_EOL;
        echo '       <form action="rascsi_action.php" method="post">'.PHP_EOL;
        echo '            <input type="hidden" name="command" value="delete_file"/>'.PHP_EOL;
        echo '            <input type="hidden" name="file_name" value="'.$file_name.'"/>'.PHP_EOL;
        echo '            <input type="submit" value="Delete">'.PHP_EOL;
        echo '       </form>'.PHP_EOL;
        echo '    </td>'.PHP_EOL;
        echo '</tr>'.PHP_EOL;
	}
?>
    </table>


    <br>
    <br>
    <h2>Upload New Image File</h2>
    <form action="rascsi_upload.php" method="post" enctype="multipart/form-data">
        <table style="border: none">
            <tr style="border: none">
                <td style="border: none; vertical-align:top;">
                    <input type="file" id="filename" name="file_name"><br><br>
                </td>
                <td style="border: none; vertical-align:top;">
                    <input type="submit" value="Upload" name="submit">
                </td>
            </tr>
        </table>
    </form>

    <br>
    <h2>Create New Empty HD Image</h2>
    <form action="rascsi_action.php" method="post">
        <input type="hidden" name="command" value="create_new_image" />
        <input type="submit" value="Create New">
    </form>

    <br>
    <h2>RaSCSI Service Status</h2>
    <table style="border: none">
            <tr style="border: none">
                <td style="border: none; vertical-align:top;">
                <form action="rascsi_action.php" method="post">
                    <input type="hidden" name="command" value="restart_rascsi_service" />
                    <input type="submit" name="restart_rascsi_service" value="Restart RaSCSI service" />
                </form>
            </td>
            <td style="border: none; vertical-align:top;">
                <form action="rascsi_action.php" method="post">
                    <input type="hidden" name="command" value="stop_rascsi_service" />
                    <input type="submit" name="stop_rascsi_service" value="Stop RaSCSI service" />
                </form>
            </td>
        </tr>
    </table>

    <br>
    <h2>Raspberry Pi Operations</h2>
    <table style="border: none">
            <tr style="border: none">
                <td style="border: none; vertical-align:top;">
                <form action="rascsi_action.php" method="post">
                    <input type="hidden" name="command" value="reboot_raspberry_pi" />
                    <input type="submit" name="reboot_rasbperry_pi" value="Reboot Raspberry Pi" />
                </form>
            </td>
            <td style="border: none; vertical-align:top;">
                <form action="rascsi_action.php" method="post">
                    <input type="hidden" name="command" value="shutdown_raspberry_pi" />
                    <input type="submit" name="shutdown_raspberry_pi" value="Shut Down Raspberry Pi" />
                </form>
            </td>
        </tr>
    </table>

</body>

</html>
